### ISSUE TITLE
gb vendor fetch: do not checkout same remote repository for different import paths

### ISSUE THREAD

**Issue by seh** [constabulary/gb#645]
When one runs _gb vendor fetch <some remote import path>_, gb calls [`main.fetch`](https://github.com/constabulary/gb/blob/master/cmd/gb-vendor/fetch.go#L84) to [acquire](https://github.com/constabulary/gb/blob/master/cmd/gb-vendor/fetch.go#L103), [copy a portion of](https://github.com/constabulary/gb/blob/master/cmd/gb-vendor/fetch.go#L134), and then [discard its copy of the remote repository](https://github.com/constabulary/gb/blob/master/cmd/gb-vendor/fetch.go#L142). After that, so long as [its `-no-recurse` flag](https://github.com/constabulary/gb/blob/master/cmd/gb-vendor/fetch.go#L40) is false, it proceeds to [fetch the missing transitive dependencies of the source it's acquired thus far](https://github.com/constabulary/gb/blob/master/cmd/gb-vendor/fetch.go#L195).

The problem arises when one requests _fetching_ of one import path from a repository that yields files that in turn import alternate paths within that same repository. Consider a hypothetical repository:
- _example.com/org/repo/.git_
- _example.com/org/repo/p1_
  - _file1.go_
```go
package p1

import "example.com/org/repo/p2"

var P p2.Something
```
- _example.com/org/repo/p2_
  - _file2.go_  
```go
package p2

type Something string
```
If one runs
```shell
gb vendor fetch example.com/org/repo/p1
```
then gb will fetch the repository _example.com/org/repo_, copy the _p1_ path within it, then proceed to fetch the same repository again, then copy the _p2_ path within it.

This doesn't matter much for small repositories, but for large ones it can take many hours, wasting bandwidth and churning the disk unnecessarily. Consider augmenting `main.fetch` to remember the set of repositories it's downloaded from its initial top-level invocation, and to destroy them all only when unwinding back up to the top-level. Intermediate recursive invocations could share that repository cache to avoid downloading the same repository more than once.

**Comment by davecheney** [comment]
Yes, this is something I need to fix. It's not just inefficient, it's
actually wrong to cherry pick parts of a repo.
